control-plane/api/v1alpha1/gatewaypolicy_types.go

```diff
@@ -0,0 +1,117 @@
+)
+
+func init() {
+	SchemeBuilder.Register(&GatewayPolicy{}, &GatewayPolicyList{})
+}
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+
+// GatewayPolicy is the Schema for the gatewaypolicies API.
+// +kubebuilder:printcolumn:name="Synced",type="string",JSONPath=".status.conditions[?(@.type==\"Synced\")].status",description="The sync status of the resource with Consul"
+// +kubebuilder:printcolumn:name="Last Synced",type="date",JSONPath=".status.lastSyncedTime",description="The last successful synced time of the resource with Consul"
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp",description="The age of the resource"
+type GatewayPolicy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   GatewayPolicySpec `json:"spec,omitempty"`
+	Status `json:"status,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+
+// GatewayPolicyList contains a list of GatewayPolicy.
+type GatewayPolicyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []GatewayPolicy `json:"items"`
+}
+
+// GatewayPolicySpec defines the desired state of GatewayPolicy.
+type GatewayPolicySpec struct {
+	// TargetRef identifies an API object to apply policy to.
+	TargetRef PolicyTargetReference `json:"targetRef"`
+	//+kubebuilder:validation:Optional
+	Override *GatewayPolicyConfig `json:"override,omitempty"`
+	//+kubebuilder:validation:Optional
+	Default *GatewayPolicyConfig `json:"default,omitempty"`
+}
+
+// PolicyTargetReference identifies the target that the policy applies to.
+type PolicyTargetReference struct {
+	// Group is the group of the target resource.
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	Group string `json:"group"`
+
+	// Kind is kind of the target resource.
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	Kind string `json:"kind"`
+
+	// Name is the name of the target resource.
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	Name string `json:"name"`
+
+	// Namespace is the namespace of the referent. When unspecified, the local
+	// namespace is inferred. Even when policy targets a resource in a different
+	// namespace, it may only apply to traffic originating from the same
+	// namespace as the policy.
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+
+	// SectionName refers to the listener targeted by this policy.
+	SectionName *gwv1beta1.SectionName `json:"sectionName,omitempty"`
+}
+
+type GatewayPolicyConfig struct {
+	//+kubebuilder:validation:Optional
+	JWT *APIGatewayJWTRequirement `json:"jwt,omitempty"`
+}
+
+// APIGatewayJWTRequirement holds the list of JWT providers to be verified against.
+type APIGatewayJWTRequirement struct {
```

Comment:
For consistency are you ok with dropping the `API` part of the naming on these structs? We're "Gateway" spec implementors and happen to call the underlying type of gateway in Consul an "API gateway" - but from the k8s perspective everything is just "Gateway" -- main reason is because these docstrings having "API" in them are part of the CRDs, so I think they'll wind up being somewhat user-facing.

Reply:
I'm okay with that, should we also drop the `API` portion of it on the consul side of things?

Reply:
RE Consul side of things, maybe, I think the only place where that'd seem a bit wonky is in the API subpackage assuming someone is creating these directly through the go client.